frame/paged-list/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+//!
+//! # Paged List Pallet
+//!
+//! A thin wrapper pallet around a [`paged_list::StoragePagedList`]. It provides an API for a single
+//! paginated list. It can be instantiated multiple times to provide multiple lists.
+//!
+//! ## Overview
+//!
+//! The pallet is quite unique since it does not expose any `Call`s, `Error`s or `Event`s. All
+//! interaction goes through the implemented [`StorageList`][frame_support::storage::StorageList]
+//! trait.
+//!
+//! A fuzzer for testing is provided in crate `pallet-paged-list-fuzzer`.
+//!
+//! ## Examples
+//!
+//! 1. **Appending** some data to the list can happen either by [`Pallet::append_one`]:
+#![doc = docify::embed!("frame/paged-list/src/tests.rs", append_one_works)]
+//! 2. or by [`Pallet::append_many`]. This should always be preferred to repeated calls to
+//! [`Pallet::append_one`]:
+#![doc = docify::embed!("frame/paged-list/src/tests.rs", append_many_works)]
+//! 3. If you want to append many values (ie. in a loop), then best use the [`Pallet::appender`]:
+#![doc = docify::embed!("frame/paged-list/src/tests.rs", appender_works)]
+//! 4. **Iterating** over the list can be done with [`Pallet::iter`]. It uses the standard
+//! `Iterator` trait:
+#![doc = docify::embed!("frame/paged-list/src/tests.rs", iter_works)]
+//! 5. **Draining** elements happens through the [`Pallet::drain`] iterator. Note that even
+//! *peeking* a value will already remove it.
+#![doc = docify::embed!("frame/paged-list/src/tests.rs", drain_works)]
+//!
+//! ## Pallet API
+//!
+//! None. Only things to consider is the [`Config`] traits.
+//!
+//! ## Low Level / Implementation Details
+//!
+//! Implementation details are documented in [`paged_list::StoragePagedList`].  
+//! All storage entries are prefixed with a unique prefix that is generated by [`ListPrefix`].
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+pub mod mock;
+mod paged_list;
+mod tests;
+
+use codec::FullCodec;
+use frame_support::{
+	pallet_prelude::StorageList,
+	traits::{PalletInfoAccess, StorageInstance},
+};
+pub use paged_list::StoragePagedList;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T, I = ()>(_);
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// The value type that can be stored in the list.
+		type Value: FullCodec;
+
+		/// The number of values that can be put into newly created pages.
+		///
+		/// Note that this does not retroactively affect already created pages. This value can be
+		/// changed at any time without requiring a runtime migration.
+		#[pallet::constant]
+		type ValuesPerNewPage: Get<u32>;
+	}
+
+	/// A storage paged list akin to what the FRAME macros would generate.
+	// Note that FRAME does natively support paged lists in storage.
+	pub type List<T, I> = StoragePagedList<
```

Comment:
This pallet only holds a single storage. It should be possible to make the PagedList struct to be generic to take storage as generic parameter so that the storage can be defined in other pallets and have no need to have an actual paged list pallet here.

This is an example https://github.com/open-web3-stack/open-runtime-module-library/blob/58146e4a227c4cc4c49669a749dc04ee54d8f009/utilities/src/linked_item.rs

Reply:
Yes, why isn't this just another storage primitive?

Reply:
Yes the pallet could provide multiple lists per instance. We can still add that. Currently it was designed in the simplest way possible to allow for some fix to be deployed in a timely manner.  

It could also have been dont as FRAME primitive, but then it needs support in the metadata.

Reply:
But now it is also not supported in metadata :P